proposals/content-layer.md

```diff
@@ -0,0 +1,345 @@
+  type: "experimental_data",
+  loader: file("src/data/dogs.json"),
+  schema: z.object({
+    id: z.string(),
+    breed: z.string(),
+    temperament: z.array(z.string()),
+  }),
+});
+
+// Custom loaders can be defined inline or imported from packages
+const podcasts = defineCollection({
+  type: "experimental_content",
+  loader: feedLoader({
+    url: "https://feeds.99percentinvisible.org/99percentinvisible",
+  }),
+  // A loader can provide its own schema, but a user-defined schema will override it.
+});
+
+export const collections = { spacecraft, dogs, podcasts };
+```
+
+This is then used in Astro pages in the same way as current content collections.
+
+# Background & Motivation
+
+Content collections are a key primitive that brings people to Astro. Content Collections make it easy to work with local content (MD, MDX, Markdoc, etc) inside of your Astro project. They give you structure (`src/content/[collection-name]/*`), schema validation for frontmatter, and querying APIs. However they are limited in a few ways:
+
+- They can only be used with local data
+- The data must be in a specific location in the project
+- Large collections can be slow to load and use a lot of memory. They can also take up a lot of space when bundling for server deployment. This places an upper limit on the number of entries that can be practically included in a collection.
+
+Content layer is designed to be a successor to content collections that addresses these limitations and opens up more use cases. It is inspired by the Gatsby data layer, but with a simpler API and no GraphQL or graph data store.
+
+# Goals
+
+- Create a successor to content collections that can be used with local and remote data.
+- Decouple content from Vite, so that data is no longer implemented as Rollup chunks.
+- Provide a simple API for defining collections with a migration path from content collections.
+- Support local files in user-defined locations with built-in file and glob loaders.
+- Initially support Markdown rendering and JSON data for local files, with support for other formats in the future.
+- Provide an API for defining custom loaders that is powerful enough to handle a wide range of use cases, including CMS integrations.
+- Provide a higher-level API for simple inline loaders.
+- Make the API scalable to tens of thousands of entries, with good performance and low memory usage.
+- Allow loaders to define their own schemas, including dynamic schemas that can be introspected from the data source.
+
+# Non-Goals
+
+- Non-goal: allowing loaders to define multiple collections automatically. e.g. separate collections would need to be manually defined for posts and categories in a blog.
+- Non-goal: dependency tracing for entries.
+- Out of scope: hot-reloading remote data.
+- Out of scope: rendering markdown from remote data. A loader could store rendered HTML, but it would be up to the loader to handle this.
+- Out of scope: custom `Content` components.
```

Comment:
```suggestion
- Allowing loaders to define multiple collections automatically. e.g. separate collections would need to be manually defined for posts and categories in a blog.
- Dependency tracing for entries.
- Hot-reloading remote data.
- Rendering markdown from remote data. A loader could store rendered HTML, but it would be up to the loader to handle this.
- Custom `Content` components.
```